source/server/guarddog_impl.cc

```diff
@@ -32,15 +32,47 @@ GuardDogImpl::GuardDogImpl(Stats::Scope& stats_scope, const Server::Configuratio
       }()),
       watchdog_miss_counter_(stats_scope.counter("server.watchdog_miss")),
       watchdog_megamiss_counter_(stats_scope.counter("server.watchdog_mega_miss")),
+      dispatcher_(api.allocateDispatcher()),
+      loop_timer_(dispatcher_->createTimer([this]() { step(); })), pending_(false),
       run_thread_(true) {
   start(api);
 }
 
+GuardDogImpl::GuardDogImpl(Stats::Scope& stats_scope, const Server::Configuration::Main& config,
+                           Api::Api& api)
+    : GuardDogImpl(stats_scope, config, api, std::make_unique<TestInterlockHook>()) {}
+
 GuardDogImpl::~GuardDogImpl() { stop(); }
 
 void GuardDogImpl::threadRoutine() {
-  do {
-    const auto now = time_source_.monotonicTime();
+  while (true) {
+    // Use a condvar to avoid spinning in Dispatcher::run() with no pending events. The
```

Comment:
I would have to page all of this code back in, but if the only reason that we have all of this condvar stuff is to avoid spinning, there are two potential fixes.
1) Just add a dummy timer which wakes up every 1 hour. This will keep it from exiting.
2) Instead of (1) you can plumb through the new EVLOOP_NO_EXIT_ON_EMPTY exit option added in recent version of libevent.

Reply:
I like (2) especially. Simplifies a bunch of things.Thanks!